### PR TITLE
Guild emblem implementation for late 2019 clients or later

### LINKED ISSUE #5114: Can't set guild emblem
* **rAthena Hash**: Newest

* **Client Date**: 20200401

* **Server Mode**: Renewal

* **Description of Issue**: When I try to choose a guild emblem it will show the list of available emblems but when I click on them they won't load / show.
  * Result: When choosing a guild emblem they won't be loaded.
  * Expected Result: Choose a guild emblem from the list clicking the edit button in the guild window and show it ingame next to the character name.
  * How to Reproduce: Create a emblem folder, insert any standard emblem 24x24 256 colors .bmp
  * Official Information: https://write.ratemyserver.net/ragnoark-online-how-to/guild-emblem/ 
screenLotusRO016
screenLotusRO017

  

* **Modifications that may affect results**:

### PR DESCRIPTION
**Pull request by secretdataz**
* **Addressed Issue(s)**: partially fixes #5114 on rAthena's side
* **Server Mode**: Irrelevant
* **Description of Pull Request**: 
Implemented guild emblem version handling on rAthena side for 2019 clients.
An external REST service is still required for hosting the guild emblems and supplying the emblem and its version to the client before it can send the actual emblem change packet to the server.

### CODE REVIEW

File: src/char/int_guild.cpp [hunk 1]
```diff
@@ -1130,6 +1130,18 @@ int mapif_guild_emblem(struct guild *g)
 	return 0;
 }
 
+// Send the guild emblem_id (version0
```
Comment (aleos89) [inline src/char/int_guild.cpp:1133]:
```suggestion
// Send the guild emblem_id (version)
```

File: src/map/clif.cpp [hunk 2]
```diff
@@ -13840,6 +13845,31 @@ void clif_parse_GuildChangeEmblem(int fd,struct map_session_data *sd){
...
+	nullpo_retv(sd);
+
+#if PACKETVER >= 20190724
+	const PACKET_CZ_GUILD_EMBLEM_CHANGE2* p = (PACKET_CZ_GUILD_EMBLEM_CHANGE2*)RFIFOP(fd, 0);
+	guild* g = sd->guild;
+
+	if (g == nullptr || g->guild_id != p->guild_id)
+		return;
+
+	int version = p->version;
```
Comment (aleos89) [inline src/map/clif.cpp:13860]: Can just do away with the assignment and just use `p->version` below as the argument for `guild_change_emblem_version`.

File: src/map/clif.cpp [hunk 3]
```diff
@@ -13840,6 +13845,31 @@ void clif_parse_GuildChangeEmblem(int fd,struct map_session_data *sd){
...
+
+	if (!sd->state.gmaster_flag)
+		return;
+
+	if (!(battle_config.emblem_woe_change) && is_agit_start()) {
```
Comment (aleos89) [inline src/map/clif.cpp:13865]:
```suggestion
	if (!battle_config.emblem_woe_change && is_agit_start()) {
```

File: src/map/guild.cpp [hunk 4]
```diff
@@ -1262,22 +1262,50 @@ int guild_notice_changed(int guild_id,const char *mes1,const char *mes2) {
...
+/*====================================================
+ * Check condition for changing guild emblem
+ *---------------------------------------------------*/
+bool guild_check_emblem_change_condition(map_session_data *sd)
+{
+	nullpo_ret(sd);
+	guild* g = sd->guild;
+
+	if (battle_config.require_glory_guild &&
+		g != nullptr && guild_checkskill(g, GD_GLORYGUILD) > 0) {
```
Comment (aleos89) [inline src/map/guild.cpp:1274]:
```suggestion
	if (battle_config.require_glory_guild && g != nullptr && guild_checkskill(g, GD_GLORYGUILD) > 0) {
```
Just in-line it. We are past our glory days of 4:3 aspect ratio.

**Comment by attackjom** [comment]
it worked.
image
image

**Comment by admkakaroto** [comment]
When I try to change the emblem I get the message:

teste_25062020

Do I need to point the config on the lua file to a specific folder? Can I use dns in the config or does it need to be an IP address?

**Comment by Balferian** [comment]
Great, it`s working.
@secretdataz when I installed athena service, tables (user_configs and guild_emblems) don`t create automaticly.
Maybe add sql tables in some folder at athena web service?

**Comment by ecdarreola** [comment]
Having problem with the PR, not sure if this is related with the web-service but I'm receiving internal server error when HTTP Post has been sent to the server. (image below)

image

**Comment by ecdarreola** [comment]
> When I try to change the emblem I get the message:
> 
> teste_25062020
> 
> Do I need to point the config on the lua file to a specific folder? Can I use dns in the config or does it need to be an IP address?

Possible cause for this is from the config/athena.php ; world name is not the same with char_server.

**Comment by Badarosk0** [comment]
> Great, it's working.
> image
> @secretdataz when I installed athena service, tables (user_configs and guild_emblems) don't create automaticly.
> Maybe add sql tables in some folder at athena web service?

is online server or localhost?

**Comment by ecdarreola** [comment]
With the web-service and this PR, emblem is successful on upload and download but emblem doesn't appear on folder.

Is it related with this PR? or just the client itself? (_tmpEmblem is not created)

**Comment by secretdataz** [comment]
@ecdarreola I’m not sure about that. I’ll merge the PR first. If the behavior you described cause a problem I’ll do a follow up later.

**Comment by voyfmyuh** [comment]
Is there a detailed tutorial? The system I use is the windows server client is 20200401

**Comment by Surefirer** [comment]
I have installed web service, got set up everything, SQL has both tables, configured .env and athena.php changed "allowed_worlds" to my server name. when I try to upload the emblem, I have the same issue says "please try again in a few mins". unless I configured something wrong, I can not figure out where I did wrong. PHP version: 7.4.0. client version: 20200318

**Comment by secretdataz** [comment]
Please note that this PR is not a support thread for the REST API.

**Comment by gravity-ro** [comment]
Can it be adjusted backward or not?
I mean before use web service.

This link it's not work
Revert "Guild emblem implementation for late 2019 clients or later

**Pull request merged**